File: CORE/src/test/java/com/variant/core/tests/TargetingTest.java
package com.variant.core.tests;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.Random;

import org.junit.Test;

import com.variant.core.VariantSession;
import com.variant.core.VariantStateRequest;
import com.variant.core.exception.VariantRuntimeException;
import com.variant.core.flashpoint.FlashpointListener;
import com.variant.core.flashpoint.TestTargetingFlashpoint;
import com.variant.core.schema.Schema;
import com.variant.core.schema.State;
import com.variant.core.schema.Test.Experience;
import com.variant.core.schema.impl.MessageTemplate;
import com.variant.core.schema.parser.ParserResponse;

public class TargetingTest extends BaseTest {

	static final int TRIALS = 1000000;
	static final float DELTA_AS_FRACTION = .025f;

	/**
	 * Basic targeting
	 * @throws Exception
	 */
	@Test
	public void basicTest() throws Exception {

		String config = 
				"{                                                             \n" +
			    "   'states':[                                                 \n" +
			    "     {  'name':'state1',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state1'                           \n" +
			    "        }                                                     \n" +
			    "     },                                                       \n" +
			    "     {  'name':'state2',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state2'                           \n" +
			    "        }                                                     \n" +
			    "     }                                                        \n" +
			    "  ],                                                          \n" +
				"  'tests':[                                                   \n" +
			    "     {                                                        \n" +
			    "        'name':'test1',                                       \n" +
			    "        'experiences':[                                       \n" +
			    "           {                                                  \n" +
			    "              'name':'A',                                     \n" +
			    "              'weight':1 ,                                    \n" +
			    "              'isControl':true                                \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'B',                                     \n" +
			    "              'weight':2                                      \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'C',                                     \n" +
			    "              'weight':97                                     \n" +
			    "           }                                                  \n" +
			    "        ],                                                    \n" +
			    "        'onStates':[                                          \n" +
			    "           {                                                  \n" +
			    "              'stateRef':'state1',                            \n" +
			    "              'variants':[                                    \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'B',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.B'       \n" +
			    "                    }                                         \n" +
			    "                 },                                           \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'C',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.C'       \n" +
			    "                    }                                         \n" +
			    "                 }                                            \n" +
			    "              ]                                               \n" +
			    "           }                                                  \n" +
			    "        ]                                                     \n" +
			    "     }                                                        \n" +
			    "  ]                                                           \n" +
			    "}                                                             \n";
		
		engine.clearFlashpointListeners();
		ParserResponse response = engine.parseSchema(config);
		if (response.hasMessages()) printMessages(response);
		assertFalse(response.hasMessages());
		Schema schema = engine.getSchema();		
		State state = schema.getState("state1");
		com.variant.core.schema.Test test = schema.getTest("test1");
		VariantSession ssn = engine.getSession("foo");

		//
		// No targeting listener - distribution according to weights.
		//
		int[] counts = {0, 0, 0};
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		verifyCounts(counts, new float[] {1, 2, 97});

		//
		// Add one null listener - still distribution according to weights.
		//
		NullTargetingFlashpointListener nullListener1 = new NullTargetingFlashpointListener();
		nullListener1.postCount = 0;
		engine.addFlashpointListener(nullListener1);
		counts[0] = counts[1] = counts[2] = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		verifyCounts(counts, new float[] {1, 2, 97});

		//
		// Add two null listeners - still distribution according to weights.
		//
		NullTargetingFlashpointListener nullListener2 = new NullTargetingFlashpointListener();
		engine.addFlashpointListener(nullListener2);
		counts[0] = counts[1] = counts[2] = 0;
		nullListener2.postCount = nullListener1.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, nullListener2.postCount);
		verifyCounts(counts, new float[] {1, 2, 97});

		//
		// Add the A/B listener - changes distribution to 1/1/0.
		//
		ABTargetingFlashpointListener ABListener = new ABTargetingFlashpointListener();
		engine.addFlashpointListener(ABListener);
		counts[0] = counts[1] = counts[2] = 0;
		ABListener.postCount = nullListener2.postCount = nullListener1.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();			
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, nullListener2.postCount);
		assertEquals(TRIALS, ABListener.postCount);
		verifyCounts(counts, new float[] {1, 1, 0});

		//
		// Clear all listeners, then add the A/C listener - changes distribution to 1/0/1.
		//
		engine.clearFlashpointListeners();
		ACTargetingFlashpointListener ACListener = new ACTargetingFlashpointListener();
		engine.addFlashpointListener(ACListener);
		counts[0] = counts[1] = counts[2] = 0;		
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {1, 0, 1});

		//
		// Add null listener - should not change the distribution.
		//
		engine.addFlashpointListener(nullListener1);
		nullListener1.postCount = ACListener.postCount = 0;
		counts[0] = counts[1] = counts[2] = 0;		
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, nullListener1.postCount);
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {1, 0, 1});

		//
		// Clear all listeners, then add the A/B/Null custom targeter + A/C targeter - changes distribution to 50/25/25.
		//
		engine.clearFlashpointListeners();
		ABNullTargetingFlashpointListener ABNullListener = new ABNullTargetingFlashpointListener();
		engine.addFlashpointListener(ABNullListener);
		engine.addFlashpointListener(ACListener);
		counts[0] = counts[1] = counts[2] = 0;
		ABNullListener.postCount = ACListener.postCount = 0;
		for (int i = 0; i < TRIALS; i++) {
			VariantStateRequest req = engine.newStateRequest(ssn, state, "");
			String expName = req.getTargetedExperience(test).getName();	
			if (expName.equals("A")) counts[0]++;
			else if (expName.equals("B")) counts[1]++;
			else if (expName.equals("C")) counts[2]++;
		} 
		assertEquals(TRIALS, ABNullListener.postCount);
		assertEquals(TRIALS, ACListener.postCount);
		verifyCounts(counts, new float[] {50, 25, 25});
	}
	
	/**
	 * 
	 * @param counts
	 * @param weights
	 */
	private static void verifyCounts(int[] counts, float[] weights) {
		float sumCounts = 0, sumWeights = 0;
		for (int i = 0; i < counts.length; i++) {
			sumCounts += counts[i];
			sumWeights += weights[i];
		}
		for (int i = 0; i < counts.length; i++) {
			//System.out.println("Delta: " + (weights[i]/sumWeights * DELTA_AS_FRACTION));
			assertEquals(weights[i]/sumWeights * DELTA_AS_FRACTION, counts[i]/sumCounts, weights[i]/sumWeights);
		}
	}

	/**
	 * Basic targeting
	 * @throws Exception
	 */
	@Test
	public void noWeightsExceptionTest() throws Exception {

		String config = 
				"{                                                             \n" +
			    "   'states':[                                                 \n" +
			    "     {  'name':'state1',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state1'                           \n" +
			    "        }                                                     \n" +
			    "     },                                                       \n" +
			    "     {  'name':'state2',                                      \n" +
			    "        'parameters':{                                        \n" +
			    "           'path':'/path/to/state2'                           \n" +
			    "        }                                                     \n" +
			    "     }                                                        \n" +
			    "  ],                                                          \n" +
				"  'tests':[                                                   \n" +
			    "     {                                                        \n" +
			    "        'name':'test1',                                       \n" +
			    "        'experiences':[                                       \n" +
			    "           {                                                  \n" +
			    "              'name':'A',                                     \n" +
			    "              'weight':1 ,                                    \n" +
			    "              'isControl':true                                \n" +
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'B'                                      \n" +
/*			    "              'weight':2                                      \n" + */
			    "           },                                                 \n" +
			    "           {                                                  \n" +
			    "              'name':'C',                                     \n" +
			    "              'weight':97                                     \n" +
			    "           }                                                  \n" +
			    "        ],                                                    \n" +
			    "        'onStates':[                                          \n" +
			    "           {                                                  \n" +
			    "              'stateRef':'state1',                            \n" +
			    "              'variants':[                                    \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'B',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.B'       \n" +
			    "                    }                                         \n" +
			    "                 },                                           \n" +
			    "                 {                                            \n" +
			    "                    'experienceRef': 'C',                     \n" +
			    "                    'parameters':{                            \n" +
			    "                       'path':'/path/to/state1/test1.C'       \n" +
			    "                    }                                         \n" +
			    "                 }                                            \n" +
			    "              ]                                               \n" +
			    "           }                                                  \n" +
			    "        ]                                                     \n" +
			    "     }                                                        \n" +
			    "  ]                                                           \n" +
			    "}                                                             \n";
		
		engine.clearFlashpointListeners();
		ParserResponse response = engine.parseSchema(config);
		if (response.hasMessages()) printMessages(response);
		assertFalse(response.hasMessages());
		Schema schema = engine.getSchema();		
		State state = schema.getState("state1");
		VariantSession ssn = engine.getSession("foo");

		boolean exceptionThrown = false;
		try {
			engine.newStateRequest(ssn, state, "");
		}
		catch (VariantRuntimeException e) {
			exceptionThrown = true;
			assertEquals(e.getMessage(), new VariantRuntimeException(MessageTemplate.RUN_WEIGHT_MISSING, "test1", "B").getMessage());
		}
		assertTrue(exceptionThrown);
	}
	
	/**
	 * Always return null.
	 */
	private static class NullTargetingFlashpointListener implements FlashpointListener<TestTargetingFlashpoint> {
		
		private int postCount = 0;
		
		@Override
		public Class<TestTargetingFlashpoint> getFlashpointClass() {
			return TestTargetingFlashpoint.class;
		}
		
		@Override
		public void post(TestTargetingFlashpoint flashpoint) {
			postCount++;
			// do nothing.
		}
	}

	/**
	 * returns A or B in equal probabilities.
	 */
	private static class ABTargetingFlashpointListener implements FlashpointListener<TestTargetingFlashpoint> {

		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());

		@Override
		public Class<TestTargetingFlashpoint> getFlashpointClass() {
			return TestTargetingFlashpoint.class;
		}
		
		@Override
		public void post(TestTargetingFlashpoint flashpoint) {
			postCount++;
			com.variant.core.schema.Test test = flashpoint.getTest();
			Experience experience = rand.nextBoolean() ? test.getExperience("A") : test.getExperience("B");
			flashpoint.setTargetedExperience(experience);
		}
	}

	/**
	 * returns A or C in equal probabilities.
	 */
	private static class ACTargetingFlashpointListener implements FlashpointListener<TestTargetingFlashpoint> {
		
		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());
		
		@Override
		public Class<TestTargetingFlashpoint> getFlashpointClass() {
			return TestTargetingFlashpoint.class;
		}
		
		@Override
		public void post(TestTargetingFlashpoint flashpoint) {
			postCount++;
			com.variant.core.schema.Test test = flashpoint.getTest();
			if (test.getName().equals("test1") && flashpoint.getTargetedExperience() == null) {
				Experience experience = rand.nextBoolean() ? test.getExperience("A") : test.getExperience("C");
				flashpoint.setTargetedExperience(experience);
			}
		}
	}

	/**
	 * returns A 25% of the time, B 25% of the time and null 50% of the time..
	 */
	private static class ABNullTargetingFlashpointListener implements FlashpointListener<TestTargetingFlashpoint> {
		
		private int postCount = 0;
		private static Random rand = new Random(System.currentTimeMillis());
		
		@Override
		public Class<TestTargetingFlashpoint> getFlashpointClass() {
			return TestTargetingFlashpoint.class;
		}
		
		@Override
		public void post(TestTargetingFlashpoint flashpoint) {
			postCount++;
			com.variant.core.schema.Test test = flashpoint.getTest();
			if (test.getName().equals("test1") && flashpoint.getTargetedExperience() == null) {
				Experience experience = rand.nextBoolean() ? (rand.nextBoolean() ? test.getExperience("A") : test.getExperience("B")) : null;
				flashpoint.setTargetedExperience(experience);
			}
		}

	}
	
	
}

File: SERVER/test/com/variant/server/test/extapi/TargetingTest.scala
package com.variant.server.test.extapi

import com.variant.server.impl.SessionImpl
import com.variant.server.test.hooks.TestTargetingHook
import com.variant.server.test.hooks.TestTargetingHookNil
import com.variant.server.test.util.ParameterizedString
import com.variant.server.api.ServerException
import com.variant.server.schema.SchemaDeployer
import com.variant.server.test.spec.BaseSpecWithServerAsync
import com.variant.core.impl.ServerError

class TargetingTest extends BaseSpecWithServerAsync {

	val trials = 500000
	val deltaAsFraction = .05f
	
   val schemaJson = """
{
   'meta':{
      'name':'${schemaName:}'
   },
   'states':[
      {
         'name':'state1'
      },
      {
         'name':'state2'
      } 
   ],
   'tests':[
      {
         'name':'test1',
         'hooks':[${hooks:}], // Defaults to null list
         'experiences':[ 
            {
               'name':'A', 
               'weight':1 , 
               'isControl':true 
            }, 
            {  
               'name':'B',
               'weight':2  
            },  
            {  
               'name':'C',
               'weight':97 
            }
         ],
         'onStates':[ 
            { 
               'stateRef':'state1',
               'variants':[ 
                  {
                     'experienceRef': 'B'
                  },
                  { 
                     'experienceRef': 'C'
                  }
               ] 
            }
         ]
      },
      {
         'name':'test2',
         'experiences':[ 
            {
               'name':'A', 
               'weight':1 , 
               'isControl':true 
            }, 
            {  
               'name':'B',
               'weight':2  
            },  
            {  
               'name':'C',
               'weight':97 
            }
         ],
         'onStates':[ 
            { 
               'stateRef':'state1',
               'variants':[ 
                  {
                     'experienceRef': 'B'
                  },
                  { 
                     'experienceRef': 'C'
                  }
               ] 
            }
         ]
      } 
   ]
}"""

   val schemaNames = Array(
         "TargetingTestNoHooks", 
         "TargetingTest1NullHook", 
         "TargetingTest2NullHooks",
         "TargetingTestABHook",
         "TargetingTestABCHook")
   
   val schemaSrc0 = ParameterizedString(schemaJson).expand(
         "schemaName" -> schemaNames(0))
   val schemaSrc1 = ParameterizedString(schemaJson).expand(
         "schemaName" -> schemaNames(1),
         "hooks" -> 
         """ {
               'name' :'nullTargetingHook',
               'class':'com.variant.server.test.hooks.TestTargetingHookNil'
             }
         """)
         
    val schemaSrc2 = ParameterizedString(schemaJson).expand(         
         "schemaName" -> schemaNames(2),
         "hooks" -> 
         """ {
               'name' :'nullTargetingHook1',
               'class':'com.variant.server.test.hooks.TestTargetingHookNil'
             },
             {
               'name' :'nullTargetingHook2',
               'class':'com.variant.server.test.hooks.TestTargetingHookNil'
             }
         """)

   val schemaSrc3 = ParameterizedString(schemaJson).expand(
         "schemaName" -> schemaNames(3),
         "hooks" -> 
         """ {
               'name' :'nullHook',
               'class':'com.variant.server.test.hooks.TestTargetingHookNil'
             },
             {
               'name' :'A_B_Hook',
               'class':'com.variant.server.test.hooks.TestTargetingHook',
               'init': {'weights':[1,1,0]}
             }
         """)
         
   val schemaSrc4 = ParameterizedString(schemaJson).expand(         
         "schemaName" -> schemaNames(4),
         "hooks" -> 
         """ {
               'name' :'nullHook',
               'class':'com.variant.server.test.hooks.TestTargetingHookNil'
             },
             {
               'name' :'A_B_CHook',
               'class':'com.variant.server.test.hooks.TestTargetingHook',
               'init': {'weights':[1,1,1]}
             }
         """)

   
	"Runtime" should {

	   "deploy multiple schemata from strings" in {
         val schemaDeployer = SchemaDeployer.fromString(schemaSrc0, schemaSrc1, schemaSrc2, schemaSrc3, schemaSrc4)
         server.useSchemaDeployer(schemaDeployer)
         schemaDeployer.parserResponses.size mustBe schemaNames.size
         schemaDeployer.parserResponses.foreach { _.getMessages.size() mustBe 0 }
         schemaNames.foreach {server.schemata.get(_).isDefined mustBe true}
	   }
	   
      "target according to weights with no targeting hooks" in {

	      async {
	         
            val schema = server.schemata.get(schemaNames(0)).get.liveGen.get
            val state = schema.getState("state1")
            val test = schema.getTest("test1")
      		
            val counts = Array(0, 0, 0)
      		for (i <- 1 to trials) {
      			val ssn = SessionImpl.empty("sid", schema)
      			ssn.targetForState(state)
      			val expName = ssn.coreSession.getStateRequest.getLiveExperience(test).getName()
      			expName match {
      			   case "A" => counts(0) += 1
         			case "B" => counts(1) += 1
         			case "C" => counts(2) += 1
      			}
      		} 
      		verifyCounts(counts, Array(1f, 2f, 97f))
	      }
      }
      
   	"target according to weights with null targeting hook" in {

	      async {

	         val schema = server.schemata.get(schemaNames(1)).get.liveGen.get
            val state = schema.getState("state1")
            val test = schema.getTest("test1")
   
      		val counts = Array(0, 0, 0)
      		for (i <- 1 to trials) {
      			val ssn = SessionImpl.empty("sid" + i, schema)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
      			val req = ssn.targetForState(state)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
      			val expName = req.getLiveExperience(test).getName()
      			expName match {
      			   case "A" => counts(0) += 1
         			case "B" => counts(1) += 1
         			case "C" => counts(2) += 1
      			}
      		} 
      		verifyCounts(counts, Array(1f, 2f, 97f))
	      }
      }
     
      "target according to weights with two null targeting hooks" in {

	      async {

            val schema = server.schemata.get(schemaNames(2)).get.liveGen.get
            val state = schema.getState("state1")
            val test = schema.getTest("test1")
   
      		val counts = Array(0, 0, 0)
      		for (i <- 1 to trials) {
      			val ssn = SessionImpl.empty("sid" + i, schema)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
      			val req = ssn.targetForState(state)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1 test1"
      			val expName = req.getLiveExperience(test).getName()
      			expName match {
      			   case "A" => counts(0) += 1
         			case "B" => counts(1) += 1
         			case "C" => counts(2) += 1
      			}
      		} 
      		verifyCounts(counts, Array(1f, 2f, 97f))
	      }
      }


      "target at 1/1/0 with the null hook and the A/B hook" in {

	      async {

            val schema = server.schemata.get(schemaNames(3)).get.liveGen.get
            val state = schema.getState("state1")
            val test = schema.getTest("test1")
   
      		val counts = Array(0, 0, 0)
      		for (i <- 1 to trials) {
      			val ssn = SessionImpl.empty("sid" + i, schema)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
      			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe null
      			val req = ssn.targetForState(state)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
      			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe "test1"
      			val expName = req.getLiveExperience(test).getName()
      			expName match {
      			   case "A" => counts(0) += 1
         			case "B" => counts(1) += 1
         			case "C" => counts(2) += 1
      			}
      		} 
      		verifyCounts(counts, Array(1f, 1f, 0f))
	      }
      }

		"still target at 1/1/1 with the null hook and the A/B/C hook" in {
         
	      async {

            val schema = server.schemata.get(schemaNames(4)).get.liveGen.get
            val state = schema.getState("state1")
            val test = schema.getTest("test1")
   
      		val counts = Array(0, 0, 0)
      		for (i <- 1 to trials) {
      			val ssn = SessionImpl.empty("sid" + i, schema)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe null
      			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe null
      			val req = ssn.targetForState(state)
      			ssn.getAttribute(TestTargetingHookNil.ATTR_KEY) mustBe "test1"
      			ssn.getAttribute(TestTargetingHook.ATTR_KEY) mustBe "test1"
      			val expName = req.getLiveExperience(test).getName()
      			expName match {
      			   case "A" => counts(0) += 1
         			case "B" => counts(1) += 1
         			case "C" => counts(2) += 1
      			}
      		} 
      		verifyCounts(counts, Array(1f, 1f, 1f))
	      }
		}
		
     "join all" in {
         joinAll(60000)  
      }
		
	   "Throw exception if targeting hook sets bad experience" in {

	      val schemaName = "bad_experience"
         val schemaSrc = ParameterizedString(schemaJson).expand(
               "schemaName" -> schemaName,
               "hooks" -> 
               """ {
                     'name' :'nullHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHookNil'
                   },
                   {
                     'name' :'A_B_CHook',
                     'class':'com.variant.server.test.hooks.TestTargetingHook',
                     'init': {'experience':'test2.A'}
                   }
               """)
               
         val schemaDeployer = SchemaDeployer.fromString(schemaSrc)
         server.useSchemaDeployer(schemaDeployer)
         val response = schemaDeployer.parserResponses(0)
         
         server.schemata.get(schemaName).isDefined mustBe true
         
         val schema = server.schemata.get(schemaName).get.liveGen.get
         val state1 = schema.getState("state1")
		   
	
		   var ssn = SessionImpl.empty(newSid(), schema)
   	   
   	   val caughtEx = intercept[ServerException.Local] {
             ssn.targetForState(state1)   // targeting hook returns an experience not from test1
         }
         caughtEx.getMessage mustBe (
                     new ServerException.Local(
                           ServerError.HOOK_TARGETING_BAD_EXPERIENCE, classOf[TestTargetingHook].getName, "test1", "test2.A"
                     ).getMessage)
 
	   }

	}
	
	/**
	 * 
	 * @param counts
	 * @param weights
	 */
	private def verifyCounts(counts: Array[Int], weights: Array[Float]) {
	   //println(counts.mkString(","))
	   //println(weights.mkString(","))
		var sumCounts = 0
		var sumWeights = 0f
		for (i <- 0 until counts.length) {
			sumCounts += counts(i)
			sumWeights += weights(i)
		}
		for (i <- 0 until counts.length) {
			//System.out.println("Delta: " + (weights[i]/sumWeights * DELTA_AS_FRACTION));
		   val countFraction = counts(i)/sumCounts.asInstanceOf[Float]
		   // +- will complain if weightsFraction is 0: if so, give it a small positive value instead
		   val weightsFraction = weights(i)/sumWeights.asInstanceOf[Float]

		   if (weightsFraction == 0)
      		countFraction mustEqual 0		      
		   else
      		countFraction mustEqual (weightsFraction +- weightsFraction * deltaAsFraction)
		}
	}
}